spark-indexer/src/main/java/org/uniprot/store/spark/indexer/uniprot/converter/UniProtEntryCommentsConverter.java

```diff
@@ -175,79 +180,95 @@ private void convertCommentAP(AlternativeProductsComment comment, UniProtDocumen
             Collection<String> values,
             Set<String> evidences,
             String experimentalField) {
-        if (hasExperimentalEvidence(evidences)) {
+        String commentVal = String.join(" ", values);
+        if (canAddExperimental(commentType, commentVal, document.reviewed, evidences)) {
             document.commentMap
                     .computeIfAbsent(experimentalField, k -> new HashSet<>())
                     .addAll(values);
+            evidences.add(EvidenceCode.Category.EXPERIMENTAL.name().toLowerCase());
+            document.evidenceExperimental = true;
         }
     }
 
-    private void convertFactor(CofactorComment comment, UniProtDocument document) {
+    private void convertFactor(
+            CofactorComment comment, UniProtDocument document, String commentVal) {
         if (comment.hasCofactors()) {
-            comment.getCofactors().forEach(cofactor -> convertCofactor(document, cofactor));
+            comment.getCofactors()
+                    .forEach(cofactor -> convertCofactor(document, commentVal, cofactor));
         }
 
         if ((comment.hasNote()) && (comment.getNote().hasTexts())) {
             List<String> noteValues = getTextsValue(comment.getNote().getTexts());
             document.cofactorNote.addAll(noteValues);
-            Set<String> textEvidences =
-                    getTextsEvidence(comment.getNote().getTexts(), document.reviewed);
-            document.cofactorNoteEv.addAll(textEvidences);
+            Set<String> textEvidences = getTextsEvidence(comment.getNote().getTexts());
             addExperimentalEvidence(
-                    document, noteValues, textEvidences, CC_COFACTOR_NOTE_EXPERIMENTAL);
+                    CommentType.COFACTOR,
+                    document,
+                    noteValues,
+                    textEvidences,
+                    CC_COFACTOR_NOTE_EXPERIMENTAL);
+            document.cofactorNoteEv.addAll(textEvidences);
         }
     }
 
-    private void convertCofactor(UniProtDocument document, Cofactor cofactor) {
+    private void convertCofactor(UniProtDocument document, String commentVal, Cofactor cofactor) {
         List<String> cofactorValues = new ArrayList<>();
         cofactorValues.add(cofactor.getName());
         if (cofactor.getCofactorCrossReference().getDatabase() == CofactorDatabase.CHEBI) {
             cofactorValues.add(cofactor.getCofactorCrossReference().getId());
         }
         document.cofactorChebi.addAll(cofactorValues);
-        if (hasExperimentalEvidence(cofactor.getEvidences())) {
+        Set<String> evidences = UniProtEntryConverterUtil.extractEvidence(cofactor.getEvidences());
+        if (canAddExperimental(CommentType.COFACTOR, commentVal, document.reviewed, evidences)) {
             document.commentMap
-                    .computeIfAbsent(CC_COFACTOR_CHEBI_EXPERIMENTAL, k -> new ArrayList<>())
+                    .computeIfAbsent(CC_COFACTOR_CHEBI_EXPERIMENTAL, k -> new HashSet<>())
```

Comment:
change from list to set. any reason for that?

Reply:
To Avoid Duplicated values